File: sdk/javascript/src/retry.js
/**
 * Retry policy for the Makra client.
 *
 * A request is retried only when repeating it is safe. GET requests always
 * are. A workflow submission is only safe because the SDK attaches an
 * `Idempotency-Key`: the gateway then replays the original run instead of
 * starting a second billable one.
 */

import { DEFAULT_RETRY_BACKOFF_MS, MAX_RETRY_BACKOFF_MS } from "./constants.js";

export const RETRYABLE_STATUS_CODES = new Set([
  408, 409, 425, 429, 500, 502, 503, 504,
]);

/**
 * 409 is retryable only for the idempotent replay race, never for a genuine
 * conflict such as reusing one key for two different bodies.
 */
const NON_RETRYABLE_CODES = new Set(["idempotency_key_reuse", "run_not_terminal"]);

export function isRetryableStatus(statusCode, code) {
  if (code && NON_RETRYABLE_CODES.has(code)) return false;
  return RETRYABLE_STATUS_CODES.has(statusCode);
}

/**
 * Milliseconds to wait before attempt `attempt` (1-based).
 *
 * A server-supplied `Retry-After` always wins. Otherwise the delay grows
 * exponentially with full jitter, which spreads out a fleet of clients that
 * were all rejected at the same moment.
 */
export function retryDelay(
  attempt,
  { backoff = DEFAULT_RETRY_BACKOFF_MS, retryAfter } = {},
) {
  if (retryAfter !== undefined && retryAfter !== null && retryAfter >= 0) {
    return Math.min(retryAfter * 1000, MAX_RETRY_BACKOFF_MS);
  }
  const ceiling = Math.min(
    backoff * 2 ** Math.max(attempt - 1, 0),
    MAX_RETRY_BACKOFF_MS,
  );
  return ceiling / 2 + Math.random() * (ceiling / 2);
}

/** Parse a `Retry-After` header expressed in seconds. */
export function parseRetryAfter(value) {
  if (!value) return undefined;
  const seconds = Number(String(value).trim());
  if (!Number.isFinite(seconds) || seconds < 0) return undefined;
  return seconds;
}

export function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: sdk/python/src/makra/_retry.py
"""Retry policy shared by the synchronous and asynchronous clients.

A request is retried only when repeating it is safe. GET requests always are.
A workflow submission is only safe because the SDK attaches an
``Idempotency-Key``: the gateway then replays the original run instead of
starting a second billable one.
"""

from __future__ import annotations

import random
from typing import Optional

from ._constants import DEFAULT_RETRY_BACKOFF, MAX_RETRY_BACKOFF

RETRYABLE_STATUS_CODES = frozenset({408, 409, 425, 429, 500, 502, 503, 504})

# 409 is retryable only for the idempotent replay race, never for a genuine
# conflict such as reusing one key for two different bodies.
_NON_RETRYABLE_CODES = frozenset({"idempotency_key_reuse", "run_not_terminal"})


def is_retryable_status(status_code: int, code: Optional[str] = None) -> bool:
    if code in _NON_RETRYABLE_CODES:
        return False
    return status_code in RETRYABLE_STATUS_CODES


def retry_delay(
    attempt: int,
    *,
    backoff: float = DEFAULT_RETRY_BACKOFF,
    retry_after: Optional[float] = None,
) -> float:
    """Seconds to wait before attempt ``attempt`` (1-based).

    A server-supplied ``Retry-After`` always wins. Otherwise the delay grows
    exponentially with full jitter, which spreads out a fleet of clients that
    were all rejected at the same moment.
    """
    if retry_after is not None and retry_after >= 0:
        return min(retry_after, MAX_RETRY_BACKOFF)
    ceiling = min(backoff * (2 ** max(attempt - 1, 0)), MAX_RETRY_BACKOFF)
    return random.uniform(ceiling / 2, ceiling)


def parse_retry_after(value: Optional[str]) -> Optional[float]:
    """Parse a ``Retry-After`` header expressed in seconds."""
    if not value:
        return None
    try:
        seconds = float(value.strip())
    except ValueError:
        return None
    return seconds if seconds >= 0 else None
